### ISSUE TITLE
Add Drag and Drop Uploads

### ISSUE THREAD

**Issue by jhubb-io** [verbb/formie#133]
It would be great to implement drag and drop uploads as standard to improve the useability of the forms.

**Comment by engram-design** [comment]
https://uppy.io/examples/dragdrop/ Seems like a feature-rich example, investigate payload size

**Comment by engram-design** [comment]
I've spent a few days on this, and not really liking the direction we're heading in terms of adding a whole lot of complexity to the submission process. There are some notable hurdles to deal with, namely around server-side validation, the form lifecycle and dealing with HTML5 file input limitations.

Notable solutions were [Uppy](https://uppy.io/) and [FilePond](https://pqina.nl/filepond/)

Firstly, We are unable to populate or modify an `<input type="file">` field. What this means in practice is that we can't use a control on the page to support drag-and-drop, and then add these items to the file input. Ideally, these files would be added to the input field, and when the form is submitted these would be sent to the server.

So, in order to get the required functionality, we need to look at some third-party solutions. The above libraries get around this by needing their own server endpoints to upload the files immediately, or by hijacking the submit of the form. While this is workable, it provides a lot of pain when we need to deal with uploading files as a separate process **just** for drag-and-drop uploads. It also poses some tricky form validation issues.

Consider this: a user fills in a form, uploads 4 images and submits the form. Only, we've set a maximum of 2 images allowed for that field. This will trigger a server-side validation error, and return to the form (if we've got "Page Reload" set in the form settings). The file upload field will now be empty, but the files have already been uploaded through the libraries server functions, that take place before the form submits. There's no way for a user to modify these uploaded files, short of adding them again, essentially overwriting them.

I'm aware client-side validation could (and does already) handle this, but we can't rely on that, as it's an opt-in setting.

This is also not mentioned being able to manage uploads. Once the form is submitted, the files are uploaded, and the user can't do anything to delete them, except replace them. Ideally, it'd be great to have a "manager" to manage uploaded files. It's currently a strange pattern to be able to show what files were uploaded to the form, but being unable to modify them at all.

As such, I'm going to close this until there is **significant** demand to implement this. There's nothing stopping you from implementing this yourself, and to your specific needs.

**Comment by engram-design** [comment]
Going to resurrect this feature request! Options are:
- [Uppy](https://uppy.io/)
- [FilePond](https://pqina.nl/filepond/)
- [Dropzone.js](https://www.dropzone.dev/js/)

**Comment by umkasanki** [comment]
Will be very useful to have this feature!

**Comment by prescottcarr** [comment]
Adding FilePond support would be amazing!

**Comment by JulieVelghe** [comment]
Is there any news on this feature request? This would be very handy to have.

**Comment by shoored** [comment]
Something like FilePond or Dropzone would be amazing! Getting this question from clients more and more.

**Comment by engram-design** [comment]
Looking into options in the next few months, maybe even a configurable provider (although not sure if that makes sense from a client perspective). The front-runner for me is Filepond due to its UI/UX being very slick.

**Comment by shoored** [comment]
Yeah Filepond looks really nice. Great that you're looking into it! 
In the meantime, I guess I can hook into the front-end templates and integrate something like this myself? 🤔

**Comment by engram-design** [comment]
You _sort of_ can, but the reason for the lack of feature is there's some work to do to get it to work. Essentially, you need to have a mechanism to upload the files separately to the submission, as they get uploaded as soon as they're selected, as opposed to now when they are uploaded submitting the form. It's totally doable (see this PR https://github.com/verbb/formie/pull/2059 using Uppy).

**Comment by shoored** [comment]
Ah okay, makes sense. Thanks!

**Comment by alexanderbuergin** [comment]
We tried to integrate FilePond into Formie in a relatively low-key way – unfortunately, this led to a number of issues. In hindsight, we clearly underestimated the migration effort involved.

It was only after deeper research that I came across your article – thanks for that! I also noticed that this feature has now been added to the milestones for version 2.3.

With that in mind: is there any rough timeline for when this feature might be implemented?

Over the past few months, we’ve received several client requests to integrate FilePond directly into Formie. In one case, we ended up using a different form solution, as integration there was more straightforward. That said, we really appreciate the depth and flexibility of Formie – which, of course, also makes integration more complex.

So the question is: what’s the current status and what are the plans for native support?

**Comment by engram-design** [comment]
@alexanderbuergin We've indeed scheduled that for 2.3, but work hasn't started on that yet. We're almost done with 2.2 (in the next month) and then we'll begin work on 2.3. As an estimate, I'd say that'll be available in the next 6 months. We're still tossing up options for the actual library to use.

We previously were going to tackle this in 2.2, but there were already [plenty](https://github.com/verbb/formie/milestone/13) of new features in that release that we wanted to ship.

**Comment by alexanderbuergin** [comment]
Thanks for your feedback.

**Comment by engram-design** [comment]
Happy to say this is in [4.0.0-beta.6](https://github.com/verbb/formie/releases/tag/4.0.0-beta.6).
